File: Daily/645. Set Mismatch/Solution4.py
from typing import List
class Solution:
    def findErrorNums(self, nums: List[int]) -> List[int]:
        '''
            Optimized Solution using Array Modification
            Time Complexity = O(n)
            Space Complexity = O(1)
        '''
        duplicate, missing = -1, 1
        for element in nums:
            if nums[abs(element) - 1] < 0: duplicate = abs(element)
            else: nums[abs(element) - 1] *= -1
        for element in range(1, len(nums)):
            if nums[element] > 0: missing = element + 1
        return [duplicate, missing]
    
if __name__ == '__main__':
    nums = [int(x) for x in input().split()]; sol = Solution()
    ans = sol.findErrorNums(nums)
    print('Duplicate Element =', ans[0], 'and Missing Element =', ans[1])

File: Daily/645. Set Mismatch/Solution4.cpp
#include <iostream>
#include <vector>
using namespace std;
class Solution {
    public:
        vector<int> findErrorNums(vector<int>& nums) {
            /*
                Optimized Solution using Array Modification
                Time Complexity = O(n)
                Space Complexity = O(1)
            */
            int duplicate = -1, missing = 1;
            const int total_Elements = nums.size();
            for(const int& element: nums)
                if(nums[abs(element) - 1] < 0) duplicate = abs(element);
                else nums[abs(element) - 1] *= -1;
            for(int element = 1; element < total_Elements; ++element)
                if(nums[element] > 0) missing = element + 1;
            return {duplicate, missing};
        }
};
int main() {
    int n; cin >> n; vector<int> nums(n); Solution sol;
    for(int index = 0; index < n; ++index) cin >> nums[index];
    vector<int> ans = sol.findErrorNums(nums);
    cout << "Duplicate Element = " << ans[0] 
        << " and Missing Element = " << ans[1] << endl;
}